File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/backend/settings/HexagonSettings.java
/*
 * Copyright 2014 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.battlelancer.seriesguide.backend.settings;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import com.battlelancer.seriesguide.provider.SeriesGuideContract;

public class HexagonSettings {

    public static final String AUDIENCE
            = "server:client_id:137959300653-9pg0ulu5d3d6jhm4fotn2onk789vsob7.apps.googleusercontent.com";

    public static final String KEY_ACCOUNT_NAME
            = "com.battlelancer.seriesguide.hexagon.v2.accountname";

    public static final String KEY_SETUP_COMPLETED
            = "com.battlelancer.seriesguide.hexagon.v2.setup_complete";

    public static final String KEY_MERGED_SHOWS
            = "com.battlelancer.seriesguide.hexagon.v2.merged.shows";

    public static final String KEY_MERGED_MOVIES
            = "com.battlelancer.seriesguide.hexagon.v2.merged.movies";

    public static final String KEY_MERGED_LISTS
            = "com.battlelancer.seriesguide.hexagon.v2.merged.lists";

    public static final String KEY_LAST_SYNC_SHOWS
            = "com.battlelancer.seriesguide.hexagon.v2.lastsync.shows";

    public static final String KEY_LAST_SYNC_EPISODES
            = "com.battlelancer.seriesguide.hexagon.v2.lastsync.episodes";

    public static final String KEY_LAST_SYNC_MOVIES
            = "com.battlelancer.seriesguide.hexagon.v2.lastsync.movies";

    public static final String KEY_LAST_SYNC_LISTS
            = "com.battlelancer.seriesguide.hexagon.v2.lastsync.lists";

    /**
     * Returns the account name used for authenticating against Hexagon, or null if not signed in.
     */
    @Nullable
    public static String getAccountName(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getString(KEY_ACCOUNT_NAME, null);
    }

    /**
     * Whether the Hexagon setup has been completed after the last sign in.
     */
    public static boolean hasCompletedSetup(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(KEY_SETUP_COMPLETED, true);
    }

    public static void setSetupCompleted(Context context) {
        setSetupState(context, true);
    }

    public static void setSetupIncomplete(Context context) {
        setSetupState(context, false);
    }

    private static void setSetupState(Context context, boolean complete) {
        PreferenceManager.getDefaultSharedPreferences(context)
                .edit().putBoolean(HexagonSettings.KEY_SETUP_COMPLETED, complete).commit();
    }

    /**
     * Reset the sync state of shows, episodes, movies and lists so a data merge is triggered when
     * next syncing with Hexagon.
     *
     * @return If the sync settings reset was committed successfully.
     */
    public static boolean resetSyncState(Context context) {
        // set all shows as not merged with Hexagon
        ContentValues values = new ContentValues();
        values.put(SeriesGuideContract.Shows.HEXAGON_MERGE_COMPLETE, false);
        context.getContentResolver()
                .update(SeriesGuideContract.Shows.CONTENT_URI, values, null, null);

        // reset sync properties
        SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(context)
                .edit();
        editor.putBoolean(HexagonSettings.KEY_MERGED_SHOWS, false);
        editor.putBoolean(HexagonSettings.KEY_MERGED_MOVIES, false);
        editor.putBoolean(HexagonSettings.KEY_MERGED_LISTS, false);
        editor.remove(HexagonSettings.KEY_LAST_SYNC_EPISODES);
        editor.remove(HexagonSettings.KEY_LAST_SYNC_SHOWS);
        editor.remove(HexagonSettings.KEY_LAST_SYNC_MOVIES);
        editor.remove(HexagonSettings.KEY_LAST_SYNC_LISTS);
        return editor.commit();
    }

    /**
     * Whether shows in the local database have been merged with those on Hexagon.
     */
    public static boolean hasMergedShows(Context context) {
        return hasMerged(context, KEY_MERGED_SHOWS);
    }

    /**
     * Whether movies in the local database have been merged with those on Hexagon.
     */
    public static boolean hasMergedMovies(Context context) {
        return hasMerged(context, KEY_MERGED_MOVIES);
    }

    /**
     * Whether lists in the local database have been merged with those on Hexagon.
     */
    public static boolean hasMergedLists(Context context) {
        return hasMerged(context, KEY_MERGED_LISTS);
    }

    private static boolean hasMerged(Context context, String key) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(key, false);
    }

    /**
     * Set the {@link #KEY_MERGED_SHOWS} setting.
     */
    public static void setHasMergedShows(Context context, boolean hasMergedShows) {
        PreferenceManager.getDefaultSharedPreferences(context)
                .edit()
                .putBoolean(KEY_MERGED_SHOWS, hasMergedShows)
                .commit();
    }

    public static long getLastEpisodesSyncTime(Context context) {
        return getLastSyncTime(context, KEY_LAST_SYNC_EPISODES);
    }

    public static long getLastShowsSyncTime(Context context) {
        return getLastSyncTime(context, KEY_LAST_SYNC_SHOWS);
    }

    public static long getLastMoviesSyncTime(Context context) {
        return getLastSyncTime(context, KEY_LAST_SYNC_MOVIES);
    }

    public static long getLastListsSyncTime(Context context) {
        return getLastSyncTime(context, KEY_LAST_SYNC_LISTS);
    }

    @SuppressLint("CommitPrefEdits")
    private static long getLastSyncTime(Context context, String key) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);

        long lastSync = prefs.getLong(key, 0);
        if (lastSync == 0) {
            lastSync = System.currentTimeMillis(); // not synced yet, then last time is now!
            prefs.edit().putLong(key, lastSync).commit();
        }

        return lastSync;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/backend/settings/HexagonSettings.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2013-2025 Uwe Trottmann

package com.battlelancer.seriesguide.backend.settings

import android.content.Context
import androidx.core.content.edit
import androidx.preference.PreferenceManager
import com.battlelancer.seriesguide.provider.SgRoomDatabase
import com.battlelancer.seriesguide.util.Utils

object HexagonSettings {

    private const val KEY_ENABLED =
        "com.battlelancer.seriesguide.hexagon.v2.enabled"
    private const val KEY_SHOULD_VALIDATE_ACCOUNT =
        "com.battlelancer.seriesguide.hexagon.v2.shouldFixAccount"
    private const val KEY_ACCOUNT_NAME =
        "com.battlelancer.seriesguide.hexagon.v2.accountname"
    private const val KEY_SETUP_COMPLETED =
        "com.battlelancer.seriesguide.hexagon.v2.setup_complete"
    private const val KEY_MERGED_SHOWS =
        "com.battlelancer.seriesguide.hexagon.v2.merged.shows"
    private const val KEY_MERGED_MOVIES =
        "com.battlelancer.seriesguide.hexagon.v2.merged.movies2"
    private const val KEY_MERGED_LISTS =
        "com.battlelancer.seriesguide.hexagon.v2.merged.lists"
    private const val KEY_LAST_SYNC_SHOWS =
        "com.battlelancer.seriesguide.hexagon.v2.lastsync.shows"
    private const val KEY_LAST_SYNC_EPISODES =
        "com.battlelancer.seriesguide.hexagon.v2.lastsync.episodes"
    private const val KEY_LAST_SYNC_MOVIES =
        "com.battlelancer.seriesguide.hexagon.v2.lastsync.movies"
    private const val KEY_LAST_SYNC_LISTS =
        "com.battlelancer.seriesguide.hexagon.v2.lastsync.lists"

    /**
     * If Cloud is enabled and Cloud specific actions should be performed or UI be shown.
     */
    fun isEnabled(context: Context): Boolean =
        Utils.hasAccessToX(context) && PreferenceManager.getDefaultSharedPreferences(context)
            .getBoolean(KEY_ENABLED, false)

    /**
     * Enable Hexagon.
     */
    fun setEnabled(context: Context) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_ENABLED, true)
            putBoolean(KEY_SHOULD_VALIDATE_ACCOUNT, false)
        }
    }

    /**
     * Disable Hexagon. Re-enabling will reset any previous sync state.
     */
    fun setDisabled(context: Context) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_ENABLED, false)
            putBoolean(KEY_SHOULD_VALIDATE_ACCOUNT, false)
        }
    }

    /**
     * Returns true if there is an issue with the current account and the user should be sent to the
     * setup screen.
     */
    fun shouldValidateAccount(context: Context): Boolean {
        return PreferenceManager.getDefaultSharedPreferences(context)
            .getBoolean(KEY_SHOULD_VALIDATE_ACCOUNT, false)
    }

    fun setShouldValidateAccount(context: Context, value: Boolean) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_SHOULD_VALIDATE_ACCOUNT, value)
        }
    }

    /**
     * Returns the account name used for authenticating against Hexagon, or null if not signed in.
     */
    fun getAccountName(context: Context): String? {
        return PreferenceManager.getDefaultSharedPreferences(context)
            .getString(KEY_ACCOUNT_NAME, null)
    }

    /**
     * Store or remove account name.
     */
    fun setAccountName(context: Context, name: String?) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putString(KEY_ACCOUNT_NAME, name)
        }
    }

    /**
     * Whether the Hexagon setup has been completed after the last sign in.
     */
    fun hasCompletedSetup(context: Context): Boolean {
        return PreferenceManager.getDefaultSharedPreferences(context)
            .getBoolean(KEY_SETUP_COMPLETED, true)
    }

    fun setSetupCompleted(context: Context) {
        setSetupState(context, true)
    }

    fun setSetupIncomplete(context: Context) {
        setSetupState(context, false)
    }

    private fun setSetupState(context: Context, complete: Boolean) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_SETUP_COMPLETED, complete)
        }
    }

    /**
     * Reset the sync state of shows, episodes, movies and lists so a data merge is triggered when
     * next syncing with Hexagon.
     *
     * @return If the sync settings reset was committed successfully.
     */
    fun resetSyncState(context: Context): Boolean {
        // set all shows as not merged with Hexagon
        SgRoomDatabase.getInstance(context).sgShow2Helper().setHexagonMergeNotCompletedForAll()

        // reset sync properties
        val editor = PreferenceManager.getDefaultSharedPreferences(context).edit()
        editor.putBoolean(KEY_MERGED_SHOWS, false)
        editor.putBoolean(KEY_MERGED_MOVIES, false)
        editor.putBoolean(KEY_MERGED_LISTS, false)
        editor.remove(KEY_LAST_SYNC_EPISODES)
        editor.remove(KEY_LAST_SYNC_SHOWS)
        editor.remove(KEY_LAST_SYNC_MOVIES)
        editor.remove(KEY_LAST_SYNC_LISTS)
        return editor.commit()
    }

    /**
     * Like [resetSyncState], but only for movies.
     */
    fun resetMovieSyncState(context: Context) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_MERGED_MOVIES, false)
            remove(KEY_LAST_SYNC_MOVIES)
        }
    }

    /**
     * Whether shows in the local database have been merged with those on Hexagon.
     */
    fun hasMergedShows(context: Context): Boolean {
        return hasMerged(context, KEY_MERGED_SHOWS)
    }

    fun setHasMergedShows(context: Context) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_MERGED_SHOWS, true)
        }
    }

    /**
     * Whether movies in the local database have been merged with those on Hexagon.
     */
    fun hasMergedMovies(context: Context): Boolean {
        return hasMerged(context, KEY_MERGED_MOVIES)
    }

    fun setHasMergedMovies(context: Context) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_MERGED_MOVIES, true)
        }
    }

    /**
     * Whether lists in the local database have been merged with those on Hexagon.
     */
    fun hasMergedLists(context: Context): Boolean {
        return hasMerged(context, KEY_MERGED_LISTS)
    }

    fun setHasMergedLists(context: Context) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(KEY_MERGED_LISTS, true)
        }
    }

    /**
     * Resets the merged state of lists so a data merge is triggered when next syncing with Hexagon.
     * Returns if the sync settings reset was committed successfully.
     */
    fun setHasNotMergedLists(context: Context): Boolean {
        return PreferenceManager.getDefaultSharedPreferences(context)
            .edit()
            .putBoolean(KEY_MERGED_LISTS, false)
            .commit()
    }

    private fun hasMerged(context: Context, key: String): Boolean {
        return PreferenceManager.getDefaultSharedPreferences(context)
            .getBoolean(key, false)
    }

    fun getLastEpisodesSyncTime(context: Context): Long {
        return getLastSyncTime(context, KEY_LAST_SYNC_EPISODES)
    }

    fun setLastEpisodesSyncTime(context: Context, timeInMs: Long) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putLong(KEY_LAST_SYNC_EPISODES, timeInMs)
        }
    }

    fun getLastShowsSyncTime(context: Context): Long {
        return getLastSyncTime(context, KEY_LAST_SYNC_SHOWS)
    }

    fun setLastShowsSyncTime(context: Context, timeInMs: Long) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putLong(KEY_LAST_SYNC_SHOWS, timeInMs)
        }
    }

    fun getLastMoviesSyncTime(context: Context): Long {
        return getLastSyncTime(context, KEY_LAST_SYNC_MOVIES)
    }

    fun setLastMoviesSyncTime(context: Context, timeInMs: Long) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putLong(KEY_LAST_SYNC_MOVIES, timeInMs)
        }
    }

    fun getLastListsSyncTime(context: Context): Long {
        return getLastSyncTime(context, KEY_LAST_SYNC_LISTS)
    }

    fun setLastListsSyncTime(context: Context, timeInMs: Long) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putLong(KEY_LAST_SYNC_LISTS, timeInMs)
        }
    }

    private fun getLastSyncTime(context: Context, key: String): Long {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        var lastSync = prefs.getLong(key, 0)
        if (lastSync == 0L) {
            lastSync = System.currentTimeMillis() // not synced yet, then last time is now!
            prefs.edit { putLong(key, lastSync) }
        }
        return lastSync
    }
}